<p>This rule raises an issue when a private method is never referenced in the code.</p>
<h2>Why is this an issue?</h2>
<p>A method that is never called is dead code, and should be removed. Cleaning out dead code decreases the size of the maintained codebase, making it
easier to understand the program and preventing bugs from being introduced.</p>
<p>This rule detects methods that are never referenced from inside a translation unit, and cannot be referenced from the outside.</p>
<h3>Code examples</h3>
<h3>Noncompliant code example</h3>
<pre data-diff-id="1" data-diff-type="noncompliant">
class Foo extends Serializable {
  private def unusedMethod(): Unit = {...} // Noncompliant
  private def writeObject(s: ObjectOutputStream): Unit = {...} // Compliant, relates to the serialization mechanism
  private def readObject(s: ObjectInputStream): Unit = {...} // Compliant, relates to the serialization mechanism
}
</pre>
<h3>Compliant solution</h3>
<pre data-diff-id="1" data-diff-type="compliant">
class Foo extends Serializable {
  private def writeObject(s: ObjectOutputStream): Unit = {...} // Compliant, relates to the serialization mechanism
  private def readObject(s: ObjectInputStream): Unit = {...} // Compliant, relates to the serialization mechanism
}
</pre>
<h3>Exceptions</h3>
<p>This rule doesnâ€™t raise issues for:</p>
<ul>
  <li> annotated methods </li>
  <li> methods with parameters that are annotated with <code>@javax.enterprise.event.Observes</code> </li>
</ul>
<p>The rule does not take reflection into account, which means that issues will be raised on <code>private</code> methods that are only accessed using
the reflection API.</p>

